# Preparing ACS PUMS Arizona Dataset for Analysis
# 1. Sourcing and loading data
## 1a. Read csv files 
Source: https://www2.census.gov/programs-surveys/acs/data/pums/2018/1-Year/

In [1]:
# Import relevant libraries 
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats
import random

In [2]:
# Read household csv file and create a dataframe 
azh = pd.read_csv('az_h.csv')
# Display dataframe to get dimensions of dataset (33533 people, 234 variables) and visualize data
azh

RT       SERIALNO  DIVISION  PUMA  REGION  ST   ADJHSG   ADJINC  WGTP  \
0      H  2018GQ0000001         8   700       4   4  1000000  1013097     0   
1      H  2018GQ0000008         8   102       4   4  1000000  1013097     0   
2      H  2018GQ0000021         8   805       4   4  1000000  1013097     0   
3      H  2018GQ0000038         8   118       4   4  1000000  1013097     0   
4      H  2018GQ0000120         8   109       4   4  1000000  1013097     0   
...   ..            ...       ...   ...     ...  ..      ...      ...   ...   
33528  H  2018HU1400736         8   118       4   4  1000000  1013097    30   
33529  H  2018HU1400743         8   203       4   4  1000000  1013097   107   
33530  H  2018HU1400818         8   126       4   4  1000000  1013097   410   
33531  H  2018HU1400876         8   122       4   4  1000000  1013097    65   
33532  H  2018HU1400883         8   101       4   4  1000000  1013097    61   

       NP  ...  WGTP71  WGTP72  WGTP73  WGTP74  WGTP75  WGTP76  WGTP77  \
0       1  ...       0       0       0       0       0       0       0   
1       1  ...       0       0       0       0       0       0       0   
2       1  ...       0       0       0       0       0       0       0   
3       1  ...       0       0       0       0       0       0       0   
4       1  ...       0       0       0       0       0       0       0   
...    ..  ...     ...     ...     ...     ...     ...     ...     ...   
33528   2  ...      34      10      31       8       8      55      55   
33529   2  ...      95     174      33      30     109     105     116   
33530   3  ...     402     467     342     104     386     728     403   
33531   1  ...      68      99      20      75      21      66     118   
33532   2  ...     102      19      63     111      17      17      60   

       WGTP78  WGTP79  WGTP80  
0           0       0       0  
1           0       0       0  
2           0       0       0  
3           0       0       0  
4           0       0       0  
...       ...     ...     ...  
33528      29      31      52  
33529     194      99     192  
33530     121     152     401  
33531      74     109      87  
33532      62      97      62  

[33533 rows x 234 columns]

In [3]:
# Do same for person csv file 
azp = pd.read_csv('az_p.csv')
azp

RT       SERIALNO  DIVISION  SPORDER  PUMA  REGION  ST   ADJINC  PWGTP  \
0      P  2018GQ0000001         8        1   700       4   4  1013097     10   
1      P  2018GQ0000008         8        1   102       4   4  1013097     25   
2      P  2018GQ0000021         8        1   805       4   4  1013097     71   
3      P  2018GQ0000038         8        1   118       4   4  1013097     26   
4      P  2018GQ0000120         8        1   109       4   4  1013097     80   
...   ..            ...       ...      ...   ...     ...  ..      ...    ...   
69985  P  2018HU1400818         8        2   126       4   4  1013097    281   
69986  P  2018HU1400818         8        3   126       4   4  1013097    336   
69987  P  2018HU1400876         8        1   122       4   4  1013097     66   
69988  P  2018HU1400883         8        1   101       4   4  1013097     61   
69989  P  2018HU1400883         8        2   101       4   4  1013097     56   

       AGEP  ...  PWGTP71  PWGTP72  PWGTP73  PWGTP74  PWGTP75  PWGTP76  \
0        52  ...       17       19        0        9       11        9   
1        78  ...        1       27       25        1       46       25   
2        28  ...      123       73       71       69       15       69   
3        24  ...       27       46       44        3        2       30   
4        19  ...       86       79       12       85       80       13   
...     ...  ...      ...      ...      ...      ...      ...      ...   
69985    34  ...      309      333      242       70      255      481   
69986     9  ...      325      442      288       81      327      551   
69987    69  ...       67       99       20       74       21       66   
69988    64  ...      102       19       63      111       18       17   
69989    59  ...       90       17       55       94       17       16   

       PWGTP77  PWGTP78  PWGTP79  PWGTP80  
0            0        9        9        0  
1           49       46       23       24  
2           16       72       64      124  
3           27       25       27       47  
4           78       84       74       80  
...        ...      ...      ...      ...  
69985      292       96      100      262  
69986      328      109      101      314  
69987      118       75      109       87  
69988       60       61       97       63  
69989       54       57       98       55  

[69990 rows x 286 columns]

## 1b. Read sas files
Ensure that these files are duplicates of the csv files and can be ignored. 

In [4]:
import pyreadstat
df, meta = pyreadstat.read_sas7bdat('az_h.sas7bdat')

# View first 3 rows 
df.head(3)

RT       SERIALNO DIVISION   PUMA REGION  ST   ADJHSG   ADJINC  WGTP   NP  \
0  H  2018GQ0000001        8  00700      4  04  1000000  1013097   0.0  1.0   
1  H  2018GQ0000008        8  00102      4  04  1000000  1013097   0.0  1.0   
2  H  2018GQ0000021        8  00805      4  04  1000000  1013097   0.0  1.0   

   ... WGTP71 WGTP72 WGTP73 WGTP74 WGTP75  WGTP76 WGTP77 WGTP78 WGTP79  WGTP80  
0  ...    0.0    0.0    0.0    0.0    0.0     0.0    0.0    0.0    0.0     0.0  
1  ...    0.0    0.0    0.0    0.0    0.0     0.0    0.0    0.0    0.0     0.0  
2  ...    0.0    0.0    0.0    0.0    0.0     0.0    0.0    0.0    0.0     0.0  

[3 rows x 234 columns]

# 2. Cleaning and transforming data

## 2a. Merge person and household files

Since we need many variables from the person files and one variable from the household files ('TEN'), we need to merge the two files, which can be done by relying on each entry's unique SERIALNO number. 

Reference doc: https://www.census.gov/programs-surveys/acs/technical-documentation/pums/filestructure.html

In [5]:
# Sort dataframes by SERIALNO
azh = azh.sort_values(by=['SERIALNO'])
azp = azp.sort_values(by=['SERIALNO'])

# Merge dataframes on SERIALNO
az = pd.merge(left=azp, right=azh, left_on='SERIALNO', right_on='SERIALNO')

# View merged dataframe and its dimensions
az

RT_x       SERIALNO  DIVISION_x  SPORDER  PUMA_x  REGION_x  ST_x  \
0        P  2018GQ0000001           8        1     700         4     4   
1        P  2018GQ0000008           8        1     102         4     4   
2        P  2018GQ0000021           8        1     805         4     4   
3        P  2018GQ0000038           8        1     118         4     4   
4        P  2018GQ0000120           8        1     109         4     4   
...    ...            ...         ...      ...     ...       ...   ...   
69985    P  2018HU1400818           8        3     126         4     4   
69986    P  2018HU1400818           8        1     126         4     4   
69987    P  2018HU1400876           8        1     122         4     4   
69988    P  2018HU1400883           8        1     101         4     4   
69989    P  2018HU1400883           8        2     101         4     4   

       ADJINC_x  PWGTP  AGEP  ...  WGTP71  WGTP72  WGTP73  WGTP74  WGTP75  \
0       1013097     10    52  ...       0       0       0       0       0   
1       1013097     25    78  ...       0       0       0       0       0   
2       1013097     71    28  ...       0       0       0       0       0   
3       1013097     26    24  ...       0       0       0       0       0   
4       1013097     80    19  ...       0       0       0       0       0   
...         ...    ...   ...  ...     ...     ...     ...     ...     ...   
69985   1013097    336     9  ...     402     467     342     104     386   
69986   1013097    410    38  ...     402     467     342     104     386   
69987   1013097     66    69  ...      68      99      20      75      21   
69988   1013097     61    64  ...     102      19      63     111      17   
69989   1013097     56    59  ...     102      19      63     111      17   

       WGTP76  WGTP77  WGTP78  WGTP79  WGTP80  
0           0       0       0       0       0  
1           0       0       0       0       0  
2           0       0       0       0       0  
3           0       0       0       0       0  
4           0       0       0       0       0  
...       ...     ...     ...     ...     ...  
69985     728     403     121     152     401  
69986     728     403     121     152     401  
69987      66     118      74     109      87  
69988      17      60      62      97      62  
69989      17      60      62      97      62  

[69990 rows x 519 columns]

## 2b. Choose relevant columns and select by industry

In [6]:
# Select columns (variables) that are relevent to our study from the merged dataframe
az = az[['INDP', 'HICOV', 'DIS', 'TEN', 'SEX', 'HISP', 'RAC1P']]
az

INDP  HICOV  DIS  TEN  SEX  HISP  RAC1P
0         NaN      1    1  NaN    1     1      1
1         NaN      1    1  NaN    1     1      1
2         NaN      2    2  NaN    1     2      1
3         NaN      1    1  NaN    1     1      1
4      8680.0      1    2  NaN    2    24      1
...       ...    ...  ...  ...  ...   ...    ...
69985     NaN      1    2  1.0    1     2      8
69986   770.0      1    2  1.0    1     2      8
69987  5391.0      1    2  3.0    1     1      2
69988     NaN      2    2  2.0    1     1      1
69989     NaN      1    2  2.0    2     1      1

[69990 rows x 7 columns]

Now I want to select only those people who work in four 'essential worker' industries. From the options in the project description, I wanted at least one industry from each of the three categories, to cover the foundation's focus areas. From there, I chose:<br><br>
1) Supermarkets and other grocery stores,<br>
2) Justice, public order, and safety activities (INDP = 9470),<br>
3) General medical and surgical hospitals (INDP = 8191), and <br>
4) Nursing care facilities (INDP = 8270).   

In [7]:
# Creat four data frames for people in the four chosen industries
grocery_store = az[az['INDP']== 4971.0]
first_responders = az[az['INDP']== 9470.0]
hospitals = az[az['INDP']== 8191.0]
nursing_facilities = az[az['INDP']== 8270.0]

# Concatenate them in order into one data frame. Now we can work on all of them or easily sort by industry 
az = grocery_store.append(first_responders).append(hospitals).append(nursing_facilities)
az

INDP  HICOV  DIS  TEN  SEX  HISP  RAC1P
618    4971.0      2    2  NaN    1     1      1
641    4971.0      1    2  NaN    1     1      1
688    4971.0      1    2  NaN    1     1      1
975    4971.0      1    2  NaN    2     1      1
1135   4971.0      1    2  NaN    2     1      1
...       ...    ...  ...  ...  ...   ...    ...
67357  8270.0      1    2  2.0    1     1      1
68206  8270.0      1    2  1.0    1     1      1
68610  8270.0      1    2  2.0    2     1      3
68648  8270.0      1    2  1.0    2     1      1
68702  8270.0      1    2  3.0    1     1      1

[3216 rows x 7 columns]

## 2c. Add 'Race' and 'HomeOwner' columns
The brief asked for a comparison between three groups categorized using two racial variables: HISP and RAC1P.  To simplify future analysis, I will create a race column to combine the two HISP and RAC1P variables with the following codes: <br>
1 = non-Hispanic white workers, HISP = 01 and RAC1P = 1<br>
2 = non-Hispanic American Indian workers, HISP = 01 and RAC1P = 03<br>
3 = Hispanic workers, HISP <> 01

In [8]:
# Define new label_race() method 
def label_race (row):
    if row['HISP'] != 1:
        return 3
    if row['HISP'] == 1:
        if row['RAC1P'] == 1:
            return 1
    if row['HISP'] == 1:
        if row['RAC1P'] == 3:
            return 2

# Use df.apply() method to fill in new 'race' column 
az['Race'] = az.apply (lambda row: label_race(row), axis=1)
# Drop old columns 
az = az.drop(['HISP', 'RAC1P'], axis=1)
az

INDP  HICOV  DIS  TEN  SEX  Race
618    4971.0      2    2  NaN    1   1.0
641    4971.0      1    2  NaN    1   1.0
688    4971.0      1    2  NaN    1   1.0
975    4971.0      1    2  NaN    2   1.0
1135   4971.0      1    2  NaN    2   1.0
...       ...    ...  ...  ...  ...   ...
67357  8270.0      1    2  2.0    1   1.0
68206  8270.0      1    2  1.0    1   1.0
68610  8270.0      1    2  2.0    2   2.0
68648  8270.0      1    2  1.0    2   1.0
68702  8270.0      1    2  3.0    1   1.0

[3216 rows x 6 columns]

The brief also asked NLC to examine whether an individual is a homeowner or renter, so I will simplify the 'TEN' column from four code answers to two: 

In [9]:
# Define new label_race() method 
def home_owner (row):
    if row['TEN'] == 1:
        return 1
    if row['TEN'] == 2:
        return 1
    if row['TEN'] == 3:
        return 2
    if row['TEN'] == 4:
        return 2

# Use df.apply() method to fill in new 'race' column 
az['HomeOwner'] = az.apply (lambda row: home_owner(row), axis=1)
# Drop old column
az = az.drop(['TEN'], axis=1)
az

INDP  HICOV  DIS  SEX  Race  HomeOwner
618    4971.0      2    2    1   1.0        NaN
641    4971.0      1    2    1   1.0        NaN
688    4971.0      1    2    1   1.0        NaN
975    4971.0      1    2    2   1.0        NaN
1135   4971.0      1    2    2   1.0        NaN
...       ...    ...  ...  ...   ...        ...
67357  8270.0      1    2    1   1.0        1.0
68206  8270.0      1    2    1   1.0        1.0
68610  8270.0      1    2    2   2.0        1.0
68648  8270.0      1    2    2   1.0        1.0
68702  8270.0      1    2    1   1.0        2.0

[3216 rows x 6 columns]

## 2d. Check data types and eliminate NaN values
Here I will use the dropna() method to elimate all NaN values, check data types, and adjust data types for readability and consistency. 

In [10]:
# Check current dimensions of data set 
az.shape

(3216, 6)

In [11]:
# Elimate NaN values (found 421)
az = az.dropna()
az.shape

(2795, 6)

In [12]:
# Check data types
az.dtypes

INDP         float64
HICOV          int64
DIS            int64
SEX            int64
Race         float64
HomeOwner    float64
dtype: object

In [13]:
# Convert float64 types to int64
az['HomeOwner'] = az['HomeOwner'].astype(int)
az['INDP'] = az['INDP'].astype(int)
az['Race'] = az['Race'].astype(int)
az

<ipython-input-13-80b9efde3e8c>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  az['HomeOwner'] = az['HomeOwner'].astype(int)
<ipython-input-13-80b9efde3e8c>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  az['INDP'] = az['INDP'].astype(int)
<ipython-input-13-80b9efde3e8c>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/u

INDP  HICOV  DIS  SEX  Race  HomeOwner
3222   4971      1    2    1     1          2
3265   4971      1    2    1     3          1
3871   4971      1    1    2     1          1
3992   4971      1    2    1     3          2
4102   4971      1    1    1     3          1
...     ...    ...  ...  ...   ...        ...
67357  8270      1    2    1     1          1
68206  8270      1    2    1     1          1
68610  8270      1    2    2     2          1
68648  8270      1    2    2     1          1
68702  8270      1    2    1     1          2

[2795 rows x 6 columns]

## 2e. Rename columns for readability

In [14]:
# View columns to ensure we are properly renaming them in a moment:
az.columns

Index(['INDP', 'HICOV', 'DIS', 'SEX', 'Race', 'HomeOwner'], dtype='object')

In [15]:
# Rename columns
az.columns = ['Industry', 'HealthInsurance', 'OccupationalDisability', 'Sex', 'Race', 'HomeOwner']
# View data frame 
az

Industry  HealthInsurance  OccupationalDisability  Sex  Race  HomeOwner
3222       4971                1                       2    1     1          2
3265       4971                1                       2    1     3          1
3871       4971                1                       1    2     1          1
3992       4971                1                       2    1     3          2
4102       4971                1                       1    1     3          1
...         ...              ...                     ...  ...   ...        ...
67357      8270                1                       2    1     1          1
68206      8270                1                       2    1     1          1
68610      8270                1                       2    2     2          1
68648      8270                1                       2    2     1          1
68702      8270                1                       2    1     1          2

[2795 rows x 6 columns]

Here we have our cleaned and transformed data set for essential workers in four industries in Arizona. Codes for variables are as follows:<br><br>
Industry: 4971=Supermarkets and grocery stores, 9470=Justice, public order, and safety activities, 8191=General medical and surgical hospitals, 8270=Nursing care facilities <br>
HealthInsurance: 1=yes, 2=no<br>
OccupationalDisability: 1=yes, 2=no <br>
Sex: 1=male, 2=female <br>
Race: 1=non-Hispanic white, 2=non-Hispanic American Indian, 3=Hispanic
HomeOwner: 1=yes, 2=no (renter) <br>